### PR TITLE
feat(dynamic-sampling): Add support for rules v2 in Sentry [TET-666]

### PR DESCRIPTION
**Pull request by iambriccardo**
This PR aims at adding support for rules v2 in sentry. This new rules format has been introduced with the goal of allow factor-based multi-matching on the Relay side.

In order to keep compatibility with old Relays, the new rules have been added to the `ruleV2` field, which is only read by newer Relays. After deployment of this PR, the plan will be to deploy the Relay changes and then followup with a PR in which the `rules` field will return `[]`. This will allow old Relays to not perform sampling and forward the sampling decisions to upstream Relays.

**Because of deployment order, the function `validate_sampling_configuration` is not up to date and thus the new `rulesV2` field is not checked. I will perform local testing before merging.**

### CODE REVIEW

File: src/sentry/dynamic_sampling/rules/utils.py [hunk 1]
```diff
@@ -144,3 +150,16 @@ def get_enabled_user_biases(user_set_biases: Optional[List[ActivatableBias]]) ->
...
+    """
+    Function responsible for defining the sampling factor for a Relay rule.
+    The input range of the function must be within [0.0, 1.0] as it is designed to make the factor decay from x to 1.
+    """
+    if base_sample_rate < 0.0 or base_sample_rate > 1.0:
+        raise Exception(
+            "The dynamic factor function requires a sample rate in the interval [0.0, 1.0]."
+        )
+
+    return float(x / (x**base_sample_rate))
```
Comment (jjbayer) [inline src/sentry/dynamic_sampling/rules/utils.py:165]: I would put the whole dynamic factor logic in a separate PR. Let's first make sure that the switch from `rules` to `rulesV2` works before we make any functional changes.

File: src/sentry/dynamic_sampling/rules/biases/boost_environments_bias.py [hunk 2]
```diff
@@ -41,10 +41,41 @@ def _generate_bias_rules(self, bias_data: BiasData) -> List[BaseRule]:
                 },
                 "active": True,
                 "id": bias_data["id"],
-            }
+            }  # type:ignore
```
Comment (jjbayer) [inline src/sentry/dynamic_sampling/rules/biases/boost_environments_bias.py:44]: Why is there a `# type: ignore` here?

File: src/sentry/dynamic_sampling/rules/biases/boost_environments_bias.py [hunk 3]
```diff
@@ -41,10 +41,41 @@ def _generate_bias_rules(self, bias_data: BiasData) -> List[BaseRule]:
...
+                            "op": "glob",
+                            "name": "trace.environment",
+                            "value": ENVIRONMENT_GLOBS,
+                            "options": {"ignoreCase": True},
+                        }
+                    ],
+                },
+                "active": True,
+                "id": bias_data["id"],
+            }  # type:ignore
```
Comment (jjbayer) [inline src/sentry/dynamic_sampling/rules/biases/boost_environments_bias.py:70]: Please add proper type hints for the new config format. Though they are not a guarantee that Relay can parse them, they help us catch obvious typos by comparing the rust types to the python types side-by-side.

**Review by jjbayer — CHANGES_REQUESTED** [review]

File: src/sentry/dynamic_sampling/rules/biases/boost_environments_bias.py (hunk 3 above)
Comment (iambriccardo) [reply src/sentry/dynamic_sampling/rules/biases/boost_environments_bias.py:70]: I did add them, you can see in the PR the `SamplingValue` in utils. The reason for `# type:ignore` is that mypy doesn't seem to like `Optional` fields. I have tried with `NotRequired` and it doesn't work.

Do you have any suggestions?

File: src/sentry/dynamic_sampling/rules/biases/boost_environments_bias.py (hunk 2 above)
Comment (iambriccardo) [reply src/sentry/dynamic_sampling/rules/biases/boost_environments_bias.py:44]: Explained in the comment below.

File: src/sentry/dynamic_sampling/rules/utils.py (hunk 1 above)
Comment (iambriccardo) [reply src/sentry/dynamic_sampling/rules/utils.py:165]: I agree, will try to update the PR by removing this.

Comment (iambriccardo) [reply src/sentry/dynamic_sampling/rules/utils.py:165]: Should we default to 5 as factor? Or just leave the sample rate?

File: src/sentry/dynamic_sampling/rules/biases/boost_environments_bias.py (hunk 3 above)
Comment (jjbayer) [reply src/sentry/dynamic_sampling/rules/biases/boost_environments_bias.py:70]: I would change the signature of this function (and all other V2 generators) to `List[RuleV2]`, and define `RuleV2` as

```python
class RuleV1(TypedDict):  # formerly known as BaseRule
   ...
   sampleRate: float

class RuleV2(TypedDict):
    ...
    samplingValue: SamplingValue
```

File: src/sentry/dynamic_sampling/rules/utils.py (hunk 1 above)
Comment (jjbayer) [reply src/sentry/dynamic_sampling/rules/utils.py:165]: I would keep the functional behavior exactly as it is on master branch, just rewritten in the new format. So I guess that means leave the sample rate as it was computed before?

File: src/sentry/dynamic_sampling/rules/biases/boost_environments_bias.py (hunk 3 above)
Comment (iambriccardo) [reply src/sentry/dynamic_sampling/rules/biases/boost_environments_bias.py:70]: Yep, will do! Looking at python fake typing it looks like there is no issue in changing types during method overriding right?

File: src/sentry/dynamic_sampling/rules/utils.py (hunk 1 above)
Comment (iambriccardo) [reply src/sentry/dynamic_sampling/rules/utils.py:165]: Perfect, will do.

File: src/sentry/dynamic_sampling/rules/biases/boost_environments_bias.py (hunk 3 above)
Comment (jjbayer) [reply src/sentry/dynamic_sampling/rules/biases/boost_environments_bias.py:70]: You could make `BiasRulesGenerator` generic over the rule version, see for example https://github.com/getsentry/sentry/blob/a9f0a281ef2534d86c54b70cc85238f60700feec/src/sentry/utils/kvstore/abstract.py#L5-L9

Comment (iambriccardo) [reply src/sentry/dynamic_sampling/rules/biases/boost_environments_bias.py]: Damn, Python type hints is its own world of trying to be something it isn't going to be.

**Comment by priscilawebdev** [comment]
instead of naming the new rules **rules_v2**, wouldn't it be better if we name the old ones **deprecated_rules**? we had a discussion about this a while ago in the Frontend TSC and this sounds similar

**Comment by iambriccardo** [comment]
> instead of naming the new rules **rules_v2**, wouldn't it be better if we name the old ones **deprecated_rules**? we had a discussion about this a while ago in the Frontend TSC and this sounds similar

Also @jjbayer proposed this. My main concern with such approach is that we might have problems if we introduce `rulesV3` because then how would we call the `rulesV2`? `legacyLegacyRule`?

I can for sure use the suggested approach, I am just not 100% convinced, maybe I am missing something.

**Comment by jjbayer** [comment]
> > instead of naming the new rules **rules_v2**, wouldn't it be better if we name the old ones **deprecated_rules**? we had a discussion about this a while ago in the Frontend TSC and this sounds similar
> 
> Also @jjbayer proposed this. My main concern with such approach is that we might have problems if we introduce `rulesV3` because then how would we call the `rulesV2`? `legacyLegacyRule`?
> 
> I can for sure use the suggested approach, I am just not 100% convinced, maybe I am missing something.

The problem here is compatibility with existing Relay instances. `rules` must contain the old format such that Relay does not reject the sampling config.

File: src/sentry/dynamic_sampling/rules/biases/boost_environments_bias.py [hunk 4]
```diff
@@ -45,6 +45,37 @@ def _generate_bias_rules(self, bias_data: BiasData) -> List[BaseRule]:
...
+                            "name": "trace.environment",
+                            "value": ENVIRONMENT_GLOBS,
+                            "options": {"ignoreCase": True},
+                        }
+                    ],
+                },
+                "active": True,
+                "id": bias_data["id"],
+            }
+        ]
```
Comment (jjbayer) [inline src/sentry/dynamic_sampling/rules/biases/boost_environments_bias.py:71]: There is a lot of code duplication in these generators, which could be prevented by calling the V1 generator from the V2 generator and then translating the result to the new format. But since this double generation will only exist for a short period of time, I think it's fine to leave it as-is.

File: src/sentry/dynamic_sampling/rules/utils.py [hunk 5]
```diff
@@ -49,6 +49,24 @@ class RuleType(Enum):
 REVERSE_RESERVED_IDS = {value: key for key, value in RESERVED_IDS.items()}
 
 
+SamplingValueType = Literal["sampleRate", "factor", "rollback"]
```
Comment (jjbayer) [inline src/sentry/dynamic_sampling/rules/utils.py:52]: What is `rollback`? Does that exist on Relay side as well?

File: src/sentry/dynamic_sampling/rules/utils.py [hunk 6]
```diff
@@ -115,6 +142,16 @@ def get_rule_hash(rule: BaseRule) -> int:
...
 
 
+def get_sampling_value(rule: PolymorphicRule) -> Optional[Tuple[str, float]]:
+    # Gets the sampling value from the rule, based on the type.
+    if "samplingValue" in rule and "value" in rule["samplingValue"]:  # type:ignore
+        return "factor", float(rule["samplingValue"]["value"])  # type:ignore
+    elif "sampleRate" in rule:
+        return "sample_rate", rule["sampleRate"]  # type:ignore
+
+    return None
```
Comment (jjbayer) [inline src/sentry/dynamic_sampling/rules/utils.py:152]: Could this be an abstract method on `BaseRule`? If you then implement it on each subtype, the `# type:ignore` should not be necessary.

**Review by jjbayer — APPROVED** [review]
Please make sure you test a complete config (all biases) against both a master Relay and a Relay with the new schema enabled.

File: src/sentry/dynamic_sampling/rules/biases/boost_environments_bias.py (hunk 4 above)
Comment (iambriccardo) [reply src/sentry/dynamic_sampling/rules/biases/boost_environments_bias.py:71]: The idea was actually to simply duplicate because we are going to fade it out.

File: src/sentry/dynamic_sampling/rules/utils.py (hunk 5 above)
Comment (iambriccardo) [reply src/sentry/dynamic_sampling/rules/utils.py:52]: Okay that is a mistake on my side, I have no idea why I wrote it. Will remove it.

File: src/sentry/dynamic_sampling/rules/utils.py (hunk 6 above)
Comment (iambriccardo) [reply src/sentry/dynamic_sampling/rules/utils.py:152]: Yep, you could do something like that.

File: src/sentry/dynamic_sampling/rules/biases/boost_environments_bias.py (hunk 4 above)
Comment (andriisoldatenko) [reply src/sentry/dynamic_sampling/rules/biases/boost_environments_bias.py:71]: agree with Joris, not sure why do we want to copy everything. It's okay for a temp solution. But it makes me think our current framework has some caveats and especially it means we can't extend easily 🤔

Comment (iambriccardo) [reply src/sentry/dynamic_sampling/rules/biases/boost_environments_bias.py:71]: It was super easy to extend, nothing difficult. The repetition has been kept just because we need both active, this is an edge case.

There are many ways of solving this problem but I prefer how explicit this solution is, especially keeping in mind that we are going to remove the old format.
